File: Planarian/Planarian/Modules/Caves/Models/CaveRevisionVm.cs
using Planarian.Model.Database.Entities.RidgeWalker;
using Planarian.Model.Database.Revisions;

namespace Planarian.Modules.Caves.Models;

public sealed record CaveRevisionListItemVm(
    string Id,
    string CaveId,
    int Sequence,
    DateTime PublishedOn,
    string? ActorUserId,
    string? ActorName,
    CaveRevisionSource Source,
    CaveRevisionOperation Operation,
    string? ChangeRequestId,
    string? ImportBatchId,
    bool IsCurrent);

public sealed record CaveRevisionHistoryVm(string CaveId, string? CurrentRevisionId,
    IReadOnlyList<CaveRevisionListItemVm> Revisions);

public sealed record CaveRevisionComparisonVm(
    CaveRevisionListItemVm Revision,
    CaveRevisionListItemVm? PreviousRevision,
    CavePublishedSnapshotV1? Previous,
    CavePublishedSnapshotV1 Current,
    CaveRevisionDiffVm? Diff);

public sealed record CaveScalarChangeVm(string Path, object? Previous, object? Current);
public sealed record CaveEntranceChangeVm(string EntranceId, IReadOnlyList<CaveScalarChangeVm> Scalars,
    IReadOnlyList<SnapshotTagReference> AddedTags, IReadOnlyList<SnapshotTagReference> RemovedTags);
public sealed record CaveFileChangeVm(string FileId, IReadOnlyList<CaveScalarChangeVm> Scalars);
public sealed record CaveLinePlotChangeVm(string LinePlotId, IReadOnlyList<CaveScalarChangeVm> Scalars);

public sealed record CaveRevisionDiffVm(
    IReadOnlyList<CaveScalarChangeVm> Scalars,
    IReadOnlyList<SnapshotTagReference> AddedTags,
    IReadOnlyList<SnapshotTagReference> RemovedTags,
    IReadOnlyList<string> AddedEntrances,
    IReadOnlyList<string> RemovedEntrances,
    IReadOnlyList<string> ChangedEntrances,
    IReadOnlyList<string> AddedFiles,
    IReadOnlyList<string> RemovedFiles,
    IReadOnlyList<string> ChangedFiles,
    IReadOnlyList<string> AddedLinePlots,
    IReadOnlyList<string> RemovedLinePlots,
    IReadOnlyList<string> ChangedLinePlots,
    IReadOnlyList<CaveEntranceChangeVm> EntranceChanges,
    IReadOnlyList<CaveFileChangeVm> FileChanges,
    IReadOnlyList<CaveLinePlotChangeVm> LinePlotChanges,
    IReadOnlyList<ReferenceMetadataChange> ReferenceMetadataChanges);

File: Planarian.Web/src/Modules/Caves/Models/CaveRevisionVm.ts
export interface CaveRevisionListItemVm {
  id: string;
  caveId: string;
  sequence: number;
  publishedOn: string;
  actorUserId?: string;
  actorName?: string;
  source: "SystemBaseline" | "ManagerEdit" | "UserSubmission" | "Import" | "System";
  operation: "Create" | "Update" | "Archive" | "Unarchive" | "Delete";
  changeRequestId?: string;
  importBatchId?: string;
  isCurrent: boolean;
}

export interface CaveRevisionHistoryVm {
  caveId: string;
  currentRevisionId?: string;
  revisions: CaveRevisionListItemVm[];
}

export interface SnapshotTagReference {
  role: string;
  tagTypeId: string;
  nameAtRevision: string;
}

export interface SnapshotReferenceVm {
  id: string;
  nameAtRevision: string;
  displayIdAtRevision?: string;
  abbreviationAtRevision?: string;
}

export interface CaveEntranceSnapshotVm {
  id: string;
  name?: string;
  isPrimary: boolean;
  description?: string;
  latitude?: number;
  longitude?: number;
  elevation?: number;
  srid: number;
  locationQualityTagId: string;
  locationQualityNameAtRevision: string;
  reportedOn?: string;
  pitDepthFeet?: number;
  tags: SnapshotTagReference[];
}

export interface CaveLinePlotSnapshotVm {
  id: string;
  name: string;
  contentHash: string;
}

export interface CaveFileSnapshotVm {
  id: string;
  fileTypeTagId: string;
  fileTypeNameAtRevision: string;
  fileName: string;
  displayName?: string;
}

export interface CaveSnapshotVm {
  caveId: string;
  accountId: string;
  name: string;
  alternateNames: string[];
  state: SnapshotReferenceVm;
  county: SnapshotReferenceVm;
  countyNumber: number;
  lengthFeet?: number | null;
  depthFeet?: number | null;
  maxPitDepthFeet?: number | null;
  numberOfPits?: number | null;
  narrative?: string;
  reportedOn?: string;
  isArchived: boolean;
  tags: SnapshotTagReference[];
  entrances: CaveEntranceSnapshotVm[];
  files: CaveFileSnapshotVm[];
  linePlots: CaveLinePlotSnapshotVm[];
}

export interface CaveScalarChangeVm {
  path: string;
  previous: unknown;
  current: unknown;
}

export interface CaveRevisionDiffVm {
  scalars: CaveScalarChangeVm[];
  addedTags: SnapshotTagReference[];
  removedTags: SnapshotTagReference[];
  addedEntrances: string[];
  removedEntrances: string[];
  changedEntrances: string[];
  addedFiles: string[];
  removedFiles: string[];
  changedFiles: string[];
  addedLinePlots: string[];
  removedLinePlots: string[];
  changedLinePlots: string[];
  entranceChanges: Array<{
    entranceId: string;
    scalars: CaveScalarChangeVm[];
    addedTags: SnapshotTagReference[];
    removedTags: SnapshotTagReference[];
  }>;
  fileChanges: Array<{
    fileId: string;
    scalars: CaveScalarChangeVm[];
  }>;
  linePlotChanges: Array<{
    linePlotId: string;
    scalars: CaveScalarChangeVm[];
  }>;
  referenceMetadataChanges: Array<{
    path: string;
    stableId: string;
    property: string;
    previousValue?: string;
    currentValue?: string;
  }>;
}

export interface CaveRevisionComparisonVm {
  revision: CaveRevisionListItemVm;
  previousRevision?: CaveRevisionListItemVm;
  previous?: CaveSnapshotVm;
  current: CaveSnapshotVm;
  diff?: CaveRevisionDiffVm;
}
